### ISSUE TITLE
Add CopyBlock/InitBlock overloads for ref byte

### ISSUE THREAD

**Issue by AtsushiKan** [dotnet/corefx#13427]
```c#
class Unsafe
{
   void CopyBlock(ref byte destination, ref byte source, uint byteCount);
   void CopyBlock(ref byte destination, ref byte source, UIntPtr byteCount);
   void CopyBlockUnaligned(ref byte destination, ref byte source, uint byteCount);
   void CopyBlockUnaligned(ref byte destination, ref byte source, UIntPtr byteCount);
   void InitBlock(ref byte destination, uint byteCount);
   void InitBlock(ref byte destination, UIntPtr byteCount);
   void InitBlockUnaligned(ref byte destination, uint byteCount);
   void InitBlockUnaligned(ref byte destination, UIntPtr byteCount);
}
```

Thin wrappers over the cpblk and initblk IL instructions to make them accessible from C#. Needed to implement Span<T>.CopyTo() efficiently.

Since this is intended to be a wrapper around the "cpblk" IL instruction, the behavior is undefined if the ranges overlap. (This suggests Span<T>.CopyTo() needs to handle three cases:

- No overlap - Use Unsafe.CopyTo<T>() - enjoy the perf.
- Overlap one way - Use forward for loop.
- Overlap the other way - Use backward for loop.

**Comment by jkotas** [comment]
@AtsushiKan Does this need to take length as well?

**Comment by jkotas** [comment]
cc @nietras @jamesqo @benaadams Thoughts about the shape of this API?

**Comment by AtsushiKan** [comment]
"Length" - yeah. For Span's purposes, an "int" (that's constrained to be non-negative) would be a natural match though it looks a little weird alongside the other CopyTo overloads with their uint and UIntPtr "byteCounts".

**Comment by benaadams** [comment]
With length it would be a bit strange, as its like array but not array, maybe something like?
```csharp
Unsafe.CopyTo<T>(ref T, ref T);
Unsafe.CopyRangeTo<T>(ref T, ref T, int count);
```

**Comment by nietras** [comment]
Off the top of my head, why should this not be called `BlockCopy` when it does the same as the existing `BlockCopy`just for `ref`s? 

At least that would be consistent with existing naming and it would also be clear this calls `cpblk`. 

I don't see any need for an overload **without** length. Or am I missing something?

**Comment by benaadams** [comment]
@nietras question is would it make sense to be length (bytecount) or count? When would you copy half a struct always starting at the start? (offset 0)

> I don't see any need for an overload without length.

Large struct copy, fast path for count == 1

**Comment by benaadams** [comment]
Hmm... actually that's just setting e.g. `ref val0 = ref val1;`

**Comment by nietras** [comment]
> Hmm... actually that's just setting e.g. ref val0 = ref val1;

Wouldn't that just copy the `ref`. Copying the value of a struct would be `val0 = val1;` And I assume that would be faster than any `cpblk` even for quite large structs.

> question is would it make sense to be length (bytecount) or count? When would you copy half a struct always starting at the start? (offset 0)

If the method would be generic with element type `T`, I agree that count should be `elementCount`. However, this would force the implementation to do `SizeOf` and multiply, when it is not needed for `byte`. So a `byte` only version seems more flexible, yet also more cumbersome. An option would be to add both a `byte` version and the generic version. 

@jkotas this depends on whether you think perf would be a problem for the generic version for `byte` with `elementCount`?

So to recap the API could be:
```csharp
void BlockCopy<T>(ref T destination, ref T source, int elementCount);
void BlockCopy(ref byte destination, ref byte source, int byteCount);
void BlockCopy(ref byte destination, ref byte source, UIntPtr byteCount);
```

Matching the version for pointers. Do we need `Unaligned` as well? Seems necessary for arbitrary `ref byte`.

**Comment by karelz** [comment]
@AtsushiKan I assume you didn't mean 1.1 which is now in deep Escrow. Are you currently working on the issue? (If not please unassign it from yourself)

**Comment by jkotas** [comment]
> this depends on whether you think perf would be a problem for the generic version for byte with elementCount?

Generic version that also handles overlapping regions will always be slower. We should have non-generic CopyBlock overloads that just wrap cpblk IL instruction and take ref byte arguments. They are natural extension of the existing CopyBlock methods that take void*. We should add these while we are on it.

Then there is the generic version, that takes number of elements and checks for overlaps. It would be nice if it has different name to make it clear that it is not functionally identical to the non-generic overload. Maybe `CopyBlock` vs. `BlockCopy<T>`? It is pretty subtle, but I cannot of anything better.

Also - since we are going to need [bool ContainsReferences<T>()](https://github.com/dotnet/coreclr/issues/1135) or similar primitive to implement the fast path through the generic version, we may want to expose it from S.R.CS as well while we are on it.

**Comment by nietras** [comment]
> Generic version that also handles overlapping regions will always be slower

I wasn't saying this should handle overlapping regions, though, just be a convenient overload. Non-overlapping generic can, of course, be implemented via the `ref byte` version of `BlockCopy` and `SizeOf`.

> Maybe CopyBlock vs. BlockCopy<T>? It is pretty subtle, but I cannot of anything better.

Wouldn't this be equivalent to [`Buffer.MemoryCopy`](https://msdn.microsoft.com/en-us/library/system.buffer.memorycopy(v=vs.110).aspx) which also does overlap checking? In fact, does this (the overlap handling version) need to be in `Unsafe`? Would it not be implemented via `Unsafe` instead, and could perhaps be put in `Buffer` instead to be consistent? Not sure why these where added to a `Buffer` class...

> `bool ContainsReferences()` or similar primitive to implement the fast path 

This would be great! Care to detail how this would be implemented in IL? 

@jkotas I saw you enabled `ref` usage via C# 7 for the `Unsafe` tests (and fixed my mistakes, my bad ;)), does this mean C# 7 can now be used in `corefx` in general? Are there any guidelines on how to do this` Visual Studio "15" Preview? I could try to implement what we have discussed, if you are okay with this, and you all don't mind perhaps waiting a little. It shouldn't take long, though.

**Comment by jkotas** [comment]
> Wouldn't this be equivalent to Buffer.MemoryCopy 

It is equivalent for structs without GC references; it is not equivalent for structs with GC references.

> In fact, does this (the overlap handling version) need to be in Unsafe?

We need this method for standalone `Span<T>` version that runs on existing runtimes. Since we cannot change Buffer on existing runtime, we have two options: Add this as public method to Unsafe; or do not add this as public method at all for now and just keep it as internal implementation detail of `Span<T>`. I think it is reasonable to have it in Unsafe class - it fits together with the overall theme of the type.

> Care to detail how this would be implemented in IL?

For the standalone Span<T> version, it would be done via reflection and result cached. We should also get JIT enlightened about Unsafe eventually and it will become compiler evaluated intrinsic at that point (in fact, the corelib internal version of Unsafe is already on that plan).

> does this mean C# 7 can now be used in corefx in general

Yes.

>  Are there any guidelines on how to do this` Visual Studio "15" Preview?

The corefx build comes with its own copy of the compiler, so you do not need to install anything to just build corefx.

To get the IDE integration for C# 7, you can try: https://dotnet.myget.org/feed/roslyn/package/vsix/eb2680f2-4e63-44a8-adf6-2e667d9f689c. This should work in VS 2015 preview. (I am not sure what works vs. does not work well with the current Visual Studio "15" preview. There is a lot of churn around installers.)

> you all don't mind perhaps waiting a little. It shouldn't take long, though.

@AtsushiKan Could you please coordinate with @nietras ?

**Comment by nietras** [comment]
> equivalent for structs without GC references; it is not equivalent for structs with GC references.

Probably wasn't very clear here, but yes, of course, the implementation needs to be different for non-blittable types.

> reasonable to have it in Unsafe class - it fits together with the overall theme of the type.

Okay, that makes sense, but this would mean it would have to be implemented in IL, right? Although as far as I can tell it is not completely necessary.

> standalone Span version, it would be done via reflection and result cached.

Would @AndreyAkinshin [BlittableHelper](https://github.com/AndreyAkinshin/BlittableStructs/blob/master/BlittableStructs/BlittableHelper.cs) be a different way to implement it? Not necessarily the best way.

Overall, I think this sounds like we should divide this in 3 parts:
 * Add `BlockCopy`/`BlockCopyUnaligned` overloads for `ref byte`.
 * Add `MemoryCopy` for `ref T` and `int elementCount` that handles overlapping regions and structs that contain or do not contain references. Can we use `Buffer.MemoryCopy` for non-reference structs? Or should the implementation be standalone?
 * Add `bool ContainsReference<T>()` via reflection, not sure best way here.

**Comment by AtsushiKan** [comment]
Blittability and containing object references are two different things. Blittable is tied to interop and how managed objects project themselves to P/Invoke.

**Comment by AtsushiKan** [comment]
ContainsReference<T> sounds like something that should go into a Reflection class - it's not inherently unsafe, and it's not something that's natural to implement in ILASM (so we'd have to figure out how to combine C# and ILASM code in S.R.C.Unsafe.) 

It could just be on System.Type... unless we want to bring S.R.TypeExtensions out of its obscurity just for this...

**Comment by jkotas** [comment]
> This would mean it would have to be implemented in IL, right? 

Yes. (Unless we want to integrate IL merge into the corefx build ... or this gets split into multiple packages.) I would probably implement it in C# first on the side, then copy&paste both the original C# commented out, and ildasmed C# version - with manual tweaks in the IL if applicable - you may be able to get a bit of extra perf by running the loop against the `ref` upper bound.

> I think this sounds like we should divide this in 3 parts

Agree.

**Comment by AtsushiKan** [comment]
Re: S.R.TypeExtensions - on second thought, probably not - we just got finished unifying that package across the platforms - the whole idea behind ContainsReference() is that Reflection is the implementation of last resort - we want to typeforward it to a runtime-specific implementation that has the inside knowledge to do the job efficiently.

**Comment by jkotas** [comment]
Also, ContainsReferences needs to be fast to be useful - making it generic over type is the most straightforward way to achieve that. If it takes the type as argument, it is slow by default and there needs to be complex pattern matching in the JIT to make it fast.

**Comment by benaadams** [comment]
ContainsReferences in a generic method `<T> where T : struct` can be branch elminated by the jit?

**Comment by jkotas** [comment]
Yes, that it would be the idea.

In the current CoreCLR internal impl, it is achieved by instantiation-specific IL. E.g. if the JIT asks for IL of `bool ContainsReferences<int>()`, the runtime gives back "return false;" that the JIT then inlines:

https://github.com/dotnet/coreclr/blob/master/src/vm/jitinterface.cpp#L6968

> `where T : struct`

I would leave it unconstrained - the struct constraint just gets into way.

**Comment by jamesqo** [comment]
Ok, I just finished reading thru this thread. From what I understand, we're adding:

- more `CopyBlock` overloads that accept `ref bytes` and do not handle overlaps
- a generic version of `CopyBlock` which does handle overlaps, and needs to have a different name
- a `ContainsReferences<T>` method which will be resolved/inlined by the JIT at compile-time

Thoughts:

- Adding new, unchecked `CopyBlock` overloads accepting refs seems like a good idea. Is there a reason why we're accepting `ref byte` here, and not `ref T`? @nietras, you mentioned earlier that

> If the method would be generic with element type T, I agree that count should be elementCount. However, this would force the implementation to do SizeOf and multiply, when it is not needed for byte.

which seems to imply you think that having only `CopyBlock<T>(ref T, ref T, int)` and not `CopyBlock(ref byte, ref byte, int)` would be slower since we would have to do an extra SizeOf/multiply. But `SizeOf<T>` is a JIT constant, and then for bytes the multiply by 1 will get optimized away, so how will calling the generic version be any slower than the byte-specific one?

- Regarding the generic not-named-`CopyBlock` we're adding that does handle overlaps, I think either of the following would be good names:
  - **`MoveBlock`**: In C, `memcpy` doesn't handle overlaps and `memmove` does. It would make sense for us to, similarly, name the checked version `MoveBlock` when the unchecked one is `CopyBlock`.
  - **`CheckedCopyBlock`**: More verbose, but also more explicit. `Unsafe` isn't going to be used all the time anyways, so the added verbosity might be ok. On the downside, people might be dissuaded from using this version since the name is so long, and use `CopyBlock` by default, leading to subtle bugs.

- `ContainsReferences`: This doesn't seem like it belongs in Unsafe. There's nothing inherently unsafe about it, it's not directly related to copying/refs, and can't be expressed naturally in IL. We should put this in a new static class, called something like `Metaprogramming`, `GenericIntrospection`, etc. (of course less verbose).

**Comment by jamesqo** [comment]
Other people who might be interested: @mikedn, @omariom, @GSPP

**Comment by jkotas** [comment]
@adamsitnik may be interested as well

**Comment by jkotas** [comment]
> ContainsReferences: This doesn't seem like it belongs in Unsafe. There's nothing inherently unsafe about it

It is in the same boat as Unsafe.SizeOf<T>() and several other existing APIs. It is not inherently unsafe, but it is only useful for low-level unsafe programming.

**Comment by nietras** [comment]
> Blittability and containing object references are two different things. Blittable is tied to interop and how managed objects project themselves to P/Invoke.

Yes, for example `boolean` is not blittable. Bad example :|. Can one say that reference types will always be non-blittable? And value types will not always be blittable? Hence a value type which contains a reference will never be blittable?

Yet, the caching scheme seems relevant. I would also think the C#/IL implementation of `ContainsReference<T>` could have a preamble that handles basic types so that these are fast even for first call e.g.:

```csharp
bool ContainsReferenceImpl<T>()
{
    if (typeof(T) == typeof(byte)) { return false; }
    if (typeof(T) == typeof(sbyte)) { return false; }
    if (typeof(T) == typeof(ushort)) { return false; }
    if (typeof(T) == typeof(short)) { return false; }
    if (typeof(T) == typeof(uint)) { return false; }
    if (typeof(T) == typeof(int)) { return false; }
    // Etc. for all the basic types, unless of course
}
```
This is a pattern seen often and that is recognized by the JIT, as far as I can tell, so this would be inlineable even for first call. That is no reflection needed. Of course, caching in a `static readonly` field should perhaps still be done even for these types.

> But SizeOf<T> is a JIT constant, and then for bytes the multiply by 1 will get optimized away, so how will calling the generic version be any slower than the byte-specific one?

Yes, I thought so too, but wanted to be sure. Also I think we have added overloads or changed method signature before due to IL complexity and JIT cost, so perhaps this is an issue as well. If not I am fine with the generic version. Can `cpblk` handle reference types or types containing references?

> MoveBlock, CheckedCopyBlock, MemoryCopy

`CheckedCopyBlock` might imply that this checks whether the copy is "safe" to do... although it is on `Unsafe`. `MoveBlock` is a good alternative, depending on whether new language is preferred from using `MemoryCopy`. Both of these are fine with me.

**Comment by omariom** [comment]
@jkotas 
> It is not inherently unsafe, but it is only useful for low-level unsafe programming.

I planned to use for something higher :smile:
https://github.com/dotnet/coreclr/issues/1135

**Comment by nietras** [comment]
Having looked at the many uses that `ContainsReference`/`IsReferenceFree` has, I don't think `Unsafe` is the best place for this. `TypeExtensions` sounds right, however I am not privy to how functionality is divided into which assemblies and why, and what dependencies can be made from the different assemblies.

I have the same issue with the overlapping memory copy. I don't think this should be in `Unsafe`. To me it looks like there is an oppurtunity here to finally consolidate the many different memory copy and clearing methods in corefx. A static class `Memory` seems suitable, although it overlaps the corefxlab `Memory<T>` work going on. Would it not make sense to try to have memory operations such as copy and clearing in one place e.g.:

```csharp
public static class Memory
{
    // Does not check for overlap
    public static void Copy(void* destination, void* source, UIntPtr byteCount);
    // Checks if T is reference free and copies accordingly
    public static void Copy<T>(ref T destination, ref T source, int elementCount);
    // plus more overloads

    // Check for overlap
    public static void CopyOverlapping(void* destination, void* source, UIntPtr byteCount);
    // Checks if T is reference free and copies accordingly
    public static void CopyOverlapping<T>(ref T destination, ref T source, int elementCount);
   
    public static void Clear(void* destination, UIntPtr byteCount);
    public static void Clear<T>(ref T destination, int elementCount);
    public static void ClearIfContainsReferences<T>(ref T destination, int elementCount);
}
```

This is not an example for a possible API. Some of these simply forward to `Unsafe`. But would this not make sense given the wide spread use of such functionality?

I can make a more final proposal in a separate issue if this seems relevant.

**Comment by jkotas** [comment]
> `TypeExtensions` sounds right

There is `TypeExtensions` class that is full of reflection APIs: https://github.com/dotnet/corefx/blob/master/src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs#L50. This does not belong togehter with reflection.

> A static class Memory seems suitable, although it overlaps the corefxlab Memory<T> work going on

Right. Memory is taken for the `Span<T>` and friends.

The existing `System.Runtime.CompilerServices.RuntimeHelpers` or `System.Buffer` types maybe better candidates for this.

**Comment by AtsushiKan** [comment]
The only qualm with S.R.TypeExtensions is that its only purpose was to "add back" desktop apis to .NetCore 1.0 using extension methods. Now that we're adding back all those apis anyway, S.R.TypeExtensions is a solution looking for a problem and we may not want people to consume it (and all those redundant extension methods) just to get ContainsReferences().

**Comment by jamesqo** [comment]
> The existing System.Runtime.CompilerServices.RuntimeHelpers or System.Buffer types maybe better candidates for this.

Agreed. I think there are too many of these types already, namely `Unsafe` in corefx, `Buffer` / `RuntimeHelpers` in coreclr, and `JitHelpers` which is internal to mscorlib. Adding another one will create more confusion.

**Comment by jamesqo** [comment]
> There is TypeExtensions class that is full of reflection APIs: https://github.com/dotnet/corefx/blob/master/src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs#L50. This does not belong together with reflection.

I mentioned this above, but would it be good to have a 'metaprogramming' class that simply contains cached values of a type's properties? For example

```cs
public static class TypeTraits<T>
{
    public static readonly bool IsValueType = typeof(T).IsValueType; // static readonlys will be jitted to consts?
    public static readonly bool IsEnum = typeof(T).IsEnum;
    ...
}

// Usage
if (TypeTraits<T>.IsEnum) { ... }
```

**Comment by nietras** [comment]
> This does not belong together with reflection.

@jkotas isn't `ContainsReference<T>` all reflection based? Why exactly do you want `ContainsReference` in `Unsafe`, if you don't mind indulging me :)?

> we may not want people to consume it (and all those redundant extension methods) just to get ContainsReferences().

Fair enough, but this is so closely related to `Type` that is seems more relevant to be put somewhere close to this. However, one could also argue its usage is mainly for memory related stuff, and that it could be added to `System.Memory.dll` instead...

> Memory is taken for the Span<T> and friends.

I would then call a `static class Memory` a "friend" ;) There is no `System.Memory` namespace, it is just an assembly name, so there should be no problem in having both a `Memory<T>` and a static class `Memory` in this assembly. The assembly could contain all the memory related functionality like copying, clearing and perhaps `ContainsReference` if this is better placed in this. At least it would be better placed here than `Unsafe` in my view.

> there are too many of these types already, namely Unsafe in corefx, Buffer / RuntimeHelpers in coreclr, and JitHelpers which is internal to mscorlib. Adding another one will create more confusion.

@jamesqo but this is my point exactly. This is scattered around in way too many places. I remember looking for memory copy functionality but never thought to look in `Buffer`... but that is pretty much set in stone now I guess. The copy and clear methods could also be added to `Buffer` at least this would then be together with existing memory copies, but I think there were some issues with this, since this would then be in coreclr?

The point raised by @omariom is that there are opportunities for optimization in all the data structures such as `List<T>` related to memory clearing (if references are contained), copying etc. but this would be good to have in a common place that can be used by other parts of `corefx`, `System.Memory` seems like this place, not the least since I assume `Span<T>` will be proliferated over the entire `corefx` API over time...

**Comment by jkotas** [comment]
> @jkotas isn't ContainsReference<T> all reflection based? 

The current implementation that tries its best for existing runtimes is internally implemented via reflection. The ultimate implementation should be part of the runtime and JIT intrinsic, like the current internal `ContainsReferences<T>()`  https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Runtime/CompilerServices/jithelpers.cs#L239

**Comment by nietras** [comment]
> ultimate implementation should be part of the runtime and JIT intrinsic

Yes, I understood that :) But does that mean `Unsafe` is the best place for the "public" API and the portable version to be? Or simply put, why on `Unsafe`? Is `Unsafe` in your view the public facing API for `JitHelpers`?

**Comment by jkotas** [comment]
`Unsafe` is a public version of the internal `JitHelpers` in my mind. Number of APIs that used to (or still are) on JitHelpers are on Unsafe as well. We should get rid of the internal `JitHelpers` and just switch to Unsafe - it was done in CoreRT already. 

Now, this does not mean that everything that is currently on JitHelpers should move to Unsafe. I agree that APIs like `ContainsReference<T>()` should rather be somewhere else, not on `Unsafe`.

**Comment by jkotas** [comment]
> 'metaprogramming' class that simply contains cached values of a type's properties?
> IsValueType 
> IsEnum

@jamesqo Do you have a good real-world examples where IsValueType or IsEnum would be used? In theory, you can take the entire Type/TypeInfo surface and duplicate it to this metaprogramming class, but it is unclear to me whether it would be of much use.

**Comment by jkotas** [comment]
> public static class Memory

These copy APIs are actually redundant with the ones exposed by `Span<T>`. I think we should just declare `Span<T>` as the public surface for this functionality.

There may be an extra validation of the source/destination buffer size, or check for overlap, but that is miniscule compared to the overall cost of the copy operation in most cases, and price worth paying for clean and secure design.

So what about this:

- Add `Clear()` method to `Span<T>`.
- Add `ContainsReferences<T>()` somewhere. My proposal would be to add it to the existing RuntimeHelpers, unless there is a good evidence that we want to have a lot similar type traits.

**Comment by nietras** [comment]
>  what about this

@jkotas I think that sounds reasonable. I, of course, have questions regarding the perf of this e.g. will `initblk`/`cpblk` use optimized functions (like `memmove`) if suitable e.g. when native memory? Or similar? 

Could there be future optimization opportunities in keeping the generic versions of `CopyBlock<T>/InitBlock<T>` which can see that step size is say 8 bytes for e.g. `double` enabling the clearing or copying to work in this step size, without having to do extra checks for <8 bytes and similar?

How should this proceed, should we make a single issue for API changes that we would like? Or one for each change? It would be good to get this all together to the FXDC so we can get a decision on this.

And thanks for the clarifications, I appreciate it.

**Comment by jkotas** [comment]
> initblk/cpblk use optimized functions

`Unsafe.CopyBlock/InitBlock` should be thin wrappers over cpblk/initblk IL instructions. They will use whatever the JIT on the given runtime decides to compile them into.

> How should this proceed

- Use this issue for `Unsafe.CopyBlock/InitBlock`
- Use https://github.com/dotnet/coreclr/issues/1135 for `ContainsReferences`
- Create a new issues for `Span<T>.Clear()`

> Could there be future optimization opportunities in keeping the generic versions of `CopyBlock<T>`/`InitBlock<T>`

Sounds potentially interesting. I think we should focus all these optimizations on `Span<T>` methods - to make `Span<T>` the best memory manipulation API that are hard to beat. We should have somebody from codegen team working on Span starting January - it will be the right time to discuss the options, or what makes sense, etc.

**Comment by nietras** [comment]
> should be thin wrappers over cpblk/initblk IL instructions

But this is only possible for non-overlapping memory, right? What about overlapping memory? For native one could forward to `memmove` via `Buffer`.

> Use this issue for Unsafe.CopyBlock/InitBlock

Yes, so we want to add one or more of the below overloads:

```csharp
void CopyBlock<T>(ref T destination, ref T source, int elementCount);
void CopyBlockUnaligned<T>(ref T destination, ref T source, int elementCount);
void InitBlock<T>(ref T destination, int elementCount);
void InitBlockUnaligned<T>(ref T destination, int elementCount);
```
AND/OR
```csharp
void CopyBlock(ref byte destination, ref byte source, uint byteCount);
void CopyBlock(ref byte destination, ref byte source, UIntPtr byteCount);
void CopyBlockUnaligned(ref byte destination, ref byte source, uint byteCount);
void CopyBlockUnaligned(ref byte destination, ref byte source, UIntPtr byteCount);
void InitBlock<T>(ref T destination, uint byteCount);
void InitBlock<T>(ref T destination, UIntPtr byteCount);
void InitBlockUnaligned<T>(ref T destination, uint byteCount);
void InitBlockUnaligned<T>(ref T destination, UIntPtr byteCount);
```
> Use dotnet/coreclr#1135 for ContainsReferences

Added comment for this https://github.com/dotnet/coreclr/issues/1135#issuecomment-262494258

> Create a new issues for Span<T>.Clear()

Created the proposal https://github.com/dotnet/corefx/issues/13915

**Comment by nietras** [comment]
We should also consider whether to add `Fill` like https://github.com/dotnet/corefx/issues/6695 to provide functionality like this on `Span<T>`, I could make an API proposal for that too, if this is of interest.

**Comment by jkotas** [comment]
> What about overlapping memory? For native one could forward to memmove via Buffer.

Yes, it is what Span<T>.Copy implementation should do, in cases where it is profitable at least. 

I am not sure about the generic InitBlock/CopyBlock overloads that take `ref T`. I am worried that they are too easy to call on something that contains GC references by accident. On the other hand, I understand that they are convenient - you do not have to cast to `byte` and/or do the pointer math.

@AtsushiKan What is your opinion on which of the above InitBlock/CopyBlock overloads to add?

**Comment by jamesqo** [comment]
@nietras, note that `Fill` should work on an element-per-element basis, not a per-byte one. Using `initblk` will only work if the sizeof T is 1, for example. I think since it cannot be easily expressed in IL, we should add it as an extension method somewhere in C#.

**Comment by AtsushiKan** [comment]
The "ref T" overrides - I'm inclined to vote against them. If we're promoting Span as the "go to" guy for memory operations, it should be the one that offers amenities like "ref T" (with a ContainsReferences check) while the Memory class offers the raw minimal mechanisms.

**Comment by nietras** [comment]
> note that Fill should work on an element-per-element basis, not a per-byte one.

@jamesqo yes I am aware of that. I wasn't saying or implying that `Fill` should forward to `initblk`. Merely that if we want `Span<T>` to be the go to implementation of memory operations, managed and unmanaged, we should ensure all the basic ops are there e.g. copying, clearing, filling and comparing, yes I would like comparing to be there too. 

Whether all functionality should be member or extension methods I am not sure, but would prefer `Fill` to be member, since it would be tied to `T` and we can do `static readonly` optimizations related to this inside `T`, but this can of course also be done outside... I would like at least a `AreEqual` or similar for `byte`, ideally for all reference free value types in some way, that might do vectorized optimizations too.

> add it as an extension method somewhere in C#.

I think it should be added to `Span<T>` so this is the place to use it from. It is also a lot more discoverable by being on Span itself.

> "ref T" overrides - I'm inclined to vote against them. If we're promoting Span as the "go to" guy for memory operations, it should be the one that offers amenities like "ref T"

Sounds like most are in favor of the non-generic version, I am fine with this, although my use case fits better with the generic version, I can just do it myself. I will probably forward to Span when this becomes available. However, will this eliminate the possibility of possible optimizations related to knowing the size of `T`? Should we ask codegen about their opinion?

**Comment by jkotas** [comment]
> However, will this eliminate the possibility of possible optimizations related to knowing the size of T?

Span-specific optimizations will know the size of `T`.

They may be possible for standalone initblk/cpblk in some cases as well, e.g. if the multiplication with size of `T` immediately precedes. But I doubt that they will matter enough for somebody to do them.

**Comment by jkotas** [comment]
I have update the proposal above with the current proposal.

**Comment by AtsushiKan** [comment]
Unsubscribing from thread as my year-end vacation is about to start. See 'yall in January.

**Comment by jkotas** [comment]
FXDC: proposal LGTM

**Comment by nietras** [comment]
Great, I will update my PR as soon as possible. Should I include the `InitBlock`* methods in same PR?

@jkotas perhaps, this issue should be renamed as well? E.g. "Add `CopyBlock/InitBlock` overloads for `ref byte`"

**Comment by AlexGhiondea** [comment]
@nietras I have assigned this issue to you.

**Comment by karelz** [comment]
API review: We need more info about the API to make the call. @jkotas let's review it with you in the room next Tuesday -- assigned to you.

**Comment by AlexRadch** [comment]
Proposal: Add FillMemory methods to fill memory with one value. It is like InitMemory but not only zero.

**Comment by jkotas** [comment]
@AlexRadch The methods on Unsafe class are meant to be thin minimal wrappers around one or a few IL instructions. There is no IL instruction that does what you are proposing, thus it is not a good candidate for Unsafe.

**Comment by terrajobst** [comment]
So the value of having `UIntPtr` is that the size if pointer sized (which in 64-bit is larger than `uint`). However, on 32-bit you don't have to pay for the widening to `long`. We considered removing the `uint` ones but those are for convenience.

Looks good as proposed.

**Comment by karelz** [comment]
@nietras assigning to you now ... let me know if you don't want to work on it.

**Comment by nietras** [comment]
@karelz it is already done, https://github.com/dotnet/corefx/pull/13848, no need for up for grabs.

**Comment by karelz** [comment]
Yep, that's why it is assigned to you.
We are not consistent in removing "up for grabs" when we assign it to someone and we don't have process to catch when a person is unassigned or doesn't deliver for a long time to put "up for grabs" back. That's why I prefer to keep it there even if it is assigned to someone ...

**Comment by jkotas** [comment]
Fixed by https://github.com/dotnet/corefx/pull/13848
